### ISSUE TITLE
The weight is from torchsharp to pytorch, and pytorch cannot set the model to eval mode during inference, otherwise the inference result will be wrong

### ISSUE THREAD

**Issue by williamlzw** [dotnet/TorchSharp#1007]
My goal is to convert the weights of torchsharp to tensorrt format, currently there is no direct conversion method. It can only be converted to pytorch weights first, then to onnx, and finally to tensorrt. However, it was found that the weight reasoning passed to pytorch set the eval mode and the reasoning was wrong. For example, Zbvs was recognized as Zhvs, but the reasoning was correct after removing the eval mode.

I use importsd.py to convert the weights. How to solve this problem? Will there be a way to convert to Onnx or tensorrt in the future?

**Comment by NiklasGustafsson** [comment]
> Will there be a way to convert to Onnx or tensorrt in the future?

There have certainly been requests made for both in the past, and there as issues open. At this point in time, we don't have enough resources on the .NET team to take on that work, but either one would be a welcome community contribution.

**Comment by williamlzw** [comment]
Does the pytorch weight for torchsharp transformations not set model.eval() when reasoning?

**Comment by NiklasGustafsson** [comment]
I'm not sure I understand the questions. TorchSharp just interfaces with libtorch, and I don't know what the default is. 

I would definitely make sure to set both `train` and `eval` explicitly for each case.

**Comment by williamlzw** [comment]
I found that torchsharp set model.eval after loading the weights. model.train must be set for correct reasoning. I suspect something is wrong with torch.save? torchshape version 0.100.0

**Comment by NiklasGustafsson** [comment]
Did it work differently in earlier versions? It didn't save the gradients before, either, and that's the difference between eval and train (the latter has gradients and data, the former just the data).

**Comment by williamlzw** [comment]
string savePath = "model.dat";
model.save(savePath);
Will the gradient be saved to dat?

**Comment by NiklasGustafsson** [comment]
No, the format that TorchSharp uses does not save the gradients. That has been so all along, so it's not something new. Gradients are only non-zero during the evaluation of an epoch (you should zero them out at the beginning of each epoch), so there should be no need to save them, ever.

If you have a scenario where you need to save modules in the middle of an epoch, with gradients, you will have to add some detail to why that is the case, so we can understand what you are trying to accomplish.

**Comment by williamlzw** [comment]
If the trained Model is loaded and set model.eval (), the reasoning result is not correct. What is the reason.

**Comment by williamlzw** [comment]
![image](https://github.com/dotnet/TorchSharp/assets/11348670/d9f74948-12d9-48cb-a175-b300ce33f551)

**Comment by williamlzw** [comment]
https://github.com/williamlzw/MicroOCR.NET

**Comment by NiklasGustafsson** [comment]
Hmmm. It works if you call `model.train()`?

**Comment by NiklasGustafsson** [comment]
I'll take a look at your repo tomorrow (time to call it a day for me).

**Comment by williamlzw** [comment]
Thank you!

**Comment by NiklasGustafsson** [comment]
I took a look at your code. It looks right to me. I'm still trying to understand what the flow is that leads to the failure. From the title, it sounds like you are exporting weights from TorchSharp and then loading them in pytorch, but the code above makes it look like you are doing the inference with TorchSharp.

Does the inference work when you run it in training mode, but not in eval mode? The one big difference, besides the collection of gradients, which should not affect the forward computation, is that Dropout layers are no-ops in eval mode. You do use Dropout, and I can imagine that with p=0.5, the difference between training and eval can be manifested in the Dropouts, especially if the dataset you've trained on is not large enough, but that is pure speculation on my part.

Did this work on versions < 0.100.0? I'm not sure if you're reporting this as something new with that version, or something that's been around for a while.

Thanks,

Niklas

**Comment by williamlzw** [comment]
public static void Infer(TrainConfig cfg)
        {
            string description = "cpu";
            if (torch.cuda.is_available())
            {
                description = $"cuda:{cfg.GpuIndex}";
            }

            var device = torch.device(description);
            var lines = File.ReadAllLines(cfg.VocabularyPath);
            string character = "";
            foreach (var line in lines)
            {
                character += line.TrimEnd();
            }
            var converter = BuildConverter(character);
            var model = BuildModel(cfg.InChannels, cfg.Nh, cfg.Depth, converter._numOfClass);
            if (cfg.ModelPath != null && cfg.ModelPath != "")
            {
                LoadModel(cfg.ModelPath, model);
            }
            model = model.to(device);
            model.eval();
            torchvision.io.DefaultImager = new torchvision.io.SkiaImager();
            var imgMat = Cv2.ImRead("G:\\dataset\\gen6\\train\\00001.jpg");
            byte[] detData = new byte[imgMat.Total() * 3];
            Marshal.Copy(imgMat.Data, detData, 0, detData.Length);
            var image = torch.tensor(detData, torch.ScalarType.Byte).reshape(imgMat.Height, imgMat.Width, 3).permute(2, 0, 1);
            //var image = torchvision.io.read_image("G:\\dataset\\gen6\\test\\00000.jpg");
            var transformOperator = torchvision.transforms.ConvertImageDType(ScalarType.Float32);
            var resizeOperator = torchvision.transforms.Resize(32, 140);
            image = transformOperator.call(image);
            image = resizeOperator.call(image).unsqueeze(0);
            image = image.to(device);
            var pred = model.call(image);
            var predList = converter.Decode(pred);
            var (predStr, predScore) = predList[0];
            Console.WriteLine(predStr);
        }

**Comment by williamlzw** [comment]
public static void Main()
            {  
                var cfg = BuildCfg();
                //TrainModel(cfg);
                Infer(cfg);
            }

**Comment by williamlzw** [comment]
I set model.eval() on a computer with only cuda11.7 installed, and used torchsharp to infer the correct result. However, when model.eval() is set on a computer with cuda11.7, cuda11.4, and cuda10.2 installed, the result of inference with torchsharp is wrong. torchsharp is all version 0.100.0.

**Comment by williamlzw** [comment]
I converted the weight of torchsharp to the weight of pytorch. The reasoning result of setting model.eval() in python code was wrong, but the reasoning result of setting model.train() was correct.

**Comment by williamlzw** [comment]
import torch
      import torch.nn as nn
      from torch.nn.functional import adaptive_avg_pool1d
      from typing import List, Tuple
      from torch import Tensor
      import numpy as np
      
      class ConvBNACT(nn.Sequential):
          def __init__(self, in_channels, out_channels, kernel_size,
                       stride=1, padding=0, groups=1):
              super().__init__(
                  nn.Conv2d(in_channels, out_channels,
                                    kernel_size, stride, padding, 1, groups, False),
                  nn.BatchNorm2d(out_channels),
                  nn.GELU()
              )
      
      
      
      class MicroBlock(nn.Module):
          def __init__(self, nh):
              super().__init__()
              self.conv1 = ConvBNACT(nh, nh, 1)
              self.conv2 = ConvBNACT(nh, nh, 3, 1, 1, nh)
      
          def forward(self, x):
              x = self.conv1(x)
              x = x + self.conv2(x)
              return x
      
      
      class MicroStage(nn.Sequential):
          def __init__(self, depth, nh):
              super().__init__(*[MicroBlock(nh) for _ in range(depth)])
      
      
      class MLP(nn.Sequential):
          def __init__(self, input_dim: int, hidden_dim: int):
              super().__init__(
                  nn.Linear(input_dim, hidden_dim),
                  nn.GELU(),
                  nn.Linear(hidden_dim, input_dim),
                  nn.Dropout(0.5)
              )
      
      
      class Residual(nn.Module):
          def __init__(self, fn: nn.Module):
              super().__init__()
              self.fn = fn
      
          def forward(self, x: torch.Tensor) -> torch.Tensor:
              return x + self.fn(x)
      
      
      class MLPBlock(nn.Module):
          def __init__(self, input_dim: int, hidden_dim: int):
              super().__init__()
              self.layerNorm1 = nn.LayerNorm(input_dim)
              self.mlp1 = MLP(input_dim, hidden_dim)
              self.layerNorm2 = nn.LayerNorm(input_dim)
              self.conv = nn.Conv2d(
                  input_dim, input_dim, kernel_size=3, padding=1, groups=input_dim, bias=False)
              self.layerNorm3 = nn.LayerNorm(input_dim)
              self.mlp2 = MLP(input_dim, hidden_dim)
      
          def forward(self, x):
              x = x.permute(0, 3, 2, 1)
              x = self.layerNorm1(x)
              x = x + self.mlp1(x)
              x = self.layerNorm2(x)
              x = x.permute(0, 3, 2, 1)
              x = self.conv(x)
              x = x.permute(0, 3, 2, 1)
              x = self.layerNorm3(x)
              x = x + self.mlp2(x)
              x = x.permute(0, 3, 2, 1)
              return x
      
      
      class MLPStage(nn.Sequential):
          def __init__(self, depth, input_dim, hidden_dim):
              super().__init__(*[MLPBlock(input_dim, hidden_dim)
                                 for _ in range(depth)])
      
      
      class Tokenizer(nn.Sequential):
          def __init__(self, in_channels: int, hidden_dim: int, out_dim: int):
              super().__init__(
                  ConvBNACT(in_channels, hidden_dim // 2, 3, 2, 1),
                  ConvBNACT(hidden_dim // 2, hidden_dim // 2, 3, 1, 1),
                  ConvBNACT(hidden_dim // 2, out_dim, 3, 1, 1),
                  nn.MaxPool2d(3, 2, 1)
              )
      
      
      class MicroMLPNet(nn.Module):
          def __init__(self, in_channels=3, nh=64, depth=2, nclass=60, img_height=32):
              super().__init__()
              """
              nh512可以
              """
              self.embed = Tokenizer(in_channels, nh, nh)
              self.microstage = MicroStage(depth, nh)
              self.mlpstages = MLPStage(depth, nh, nh)
              self.flatten = nn.Flatten(1, 2)
              self.dropout = nn.Dropout(0.5)
              linear_in = nh * int(img_height//4)
              self.fc = nn.Linear(linear_in, nclass)
      
          def forward(self, x):
              x_shape = x.size()
              x = self.embed(x)
              x = self.microstage(x)
              x = self.mlpstages(x)
              x = self.flatten(x)
              x = adaptive_avg_pool1d(x, int(x_shape[3]/4))
              x = x.permute(0, 2, 1)
              x = self.dropout(x)
              x = self.fc(x)
              return x
      
      class CTCLabelConverter(object):
          def __init__(self, character: str):
              list_character = list(character)
              self.num_of_classes = len(character)+2
              self.idx2char = []
              self.idx2char.append('_')
              for line in list_character:
                  line = line.strip()
                  if line != '':
                      self.idx2char.append(line)
              self.idx2char.append(' ')
              self.char2idx = {}
              for idx, char in enumerate(self.idx2char):
                  self.char2idx[char] = idx
      
          def str2idx(self, strings):
              """Convert strings to indexes.
              Args:
                  strings (list[str]): ['hello', 'world'].
              Returns:
                  indexes (list[list[int]]): [1,2,3,3,4,5,4,6,3,7].
              """
              indexes = []
              for string in strings:
                  for char in string:
                      char_idx = self.char2idx.get(char)
                      if char_idx is None:
                          raise Exception(f'Chararcter: {char} not in dict')
                      indexes.append(char_idx)
              return indexes
      
          def encode(self, strings: List[str]) -> Tuple[Tensor, Tensor]:
              targets_lengths = [len(s) for s in strings]
              targets = self.str2idx(strings)
              return torch.LongTensor(targets), torch.LongTensor(targets_lengths)
      
          def decode(self,
                     preds: Tensor,
                     raw: bool = False) -> List[Tuple[str, np.ndarray]]:
              preds = preds.softmax(dim=2)
              preds_score, preds_idx = preds.max(dim=2)
              preds_idx = preds_idx.detach().cpu().numpy().tolist()
              preds_score = preds_score.detach().cpu().numpy().tolist()
              result_list = []
              for word, score in zip(preds_idx, preds_score):
                  if raw:
                      result_list.append(
                          (''.join([self.idx2char[char_idx] for char_idx in word]), score))
                  else:
                      char_list = []
                      score_list = []
                      for i, char_idx in enumerate(word):
                          if char_idx != 0 and (not (i > 0 and word[i - 1] == char_idx)):
                              char_list.append(self.idx2char[char_idx])
                              score_list.append(score[i])
                      result_list.append((''.join(char_list), score_list))
              return result_list
      
      if __name__ == '__main__':
          import importsd
          from torchvision import transforms
          import cv2
      
          model = MicroMLPNet(3, 64, 2, 64)
          f = open("model_weights.dat", "rb") #model_weights.dat from torchsharp
          static_dict = importsd.load_state_dict(f)
          model.load_state_dict(static_dict)
          f.close()
      
          img = cv2.imread('G:/dataset/gen6/train/00000.jpg')
      
          tensor = transforms.ToTensor()(img)
          tensor = transforms.Resize((32, 140))(tensor)
          tensor = tensor.unsqueeze(dim=0)
          #model.eval() #no correct !!!!!!!!!!!!!!!!!!!!
          out = model(tensor)
      
          character = []
          vocabulary_path = './keys/english.txt'
          with open(vocabulary_path, mode='r', encoding='utf-8') as fa:
              lines = fa.readlines()
              for line in lines:
                  character.append(line.strip())
          converter = CTCLabelConverter(character)
          txt = converter.decode(out, False)
      
          print(txt)

**Comment by williamlzw** [comment]
[test.zip](https://github.com/dotnet/TorchSharp/files/11779190/test.zip)
It is caused by my problem, torchsharp weights can be loaded by python for reasoning, with less loss of precision.
